### PR TITLE
Refactor and extend statistics module

### PR DESCRIPTION
**Pull request by isabelge**
This PR integrates all statistics into the monitoring module. 

The workload monitoring can now be used to actively track changes to build statistics. Instead of SQL queries, the passive tracking now builds nodes in order to gather information.

Statistic changes:
- active tracking of unique values, min, max and row count
- DML tracking on tables (INSERT, DELETE, UPDATE, SELECT)
- tracking of commits and rollbacks
- added possibility to reevaluate row count

additional changes:
- integrated table row count
- added TableID to Table

### CODE REVIEW

**Comment by hennlo** [comment]
@isabelge I'm just curious what was the cause of the heap space error?

**Comment by isabelge** [comment]
> @isabelge I'm just curious what was the cause of the heap space error?
@hennlo The issue was the debug messages from the CatalogInfoPage. These are now limited to the last 100 messages.

File: core/src/main/java/org/polypheny/db/StatisticsManager.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+public abstract class StatisticsManager<T extends Comparable<T>> implements PropertyChangeListener {
+
+    public static StatisticsManager<?> INSTANCE = null;
+
+
+    public static StatisticsManager<?> setAndGetInstance( StatisticsManager transaction ) {
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/StatisticsManager.java:29]: Why is this parameter called transaction?

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/DdlEvent.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+package org.polypheny.db.monitoring.events;
+
+import java.util.Arrays;
+import java.util.List;
+import org.polypheny.db.monitoring.events.analyzer.DdlEventAnalyzer;
+import org.polypheny.db.monitoring.events.metrics.DdlDataPoint;
+
+public class DdlEvent extends StatementEvent {
+
+    private String eventType = "DDL EVENT";
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/DdlEvent.java:26]: Can this be static final EVENT_TYPE? However, it is never used

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/DmlDataPoint.java [hunk 3]
```diff
@@ -49,11 +50,14 @@
...
     private long executionTime;
     private boolean isSubQuery;
+    protected boolean isCommitted;
     private int rowCount;
     private List<String> fieldNames;
     private List<Long> accessedPartitions;
     private String queryClass;
     private String physicalQueryClass;
+    protected HashMap<Long, List<Object>> changedValues;
+    protected Map<Long, Long> availableColumnsWithTable;
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/DmlDataPoint.java:60]: Can these be final?

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/QueryDataPoint.java [hunk 4]
```diff
@@ -48,8 +49,11 @@
     private String description;
     private long executionTime;
     private boolean isSubQuery;
+    protected boolean isCommitted;
     private int rowCount;
     private List<String> fieldNames;
+    protected Integer indexSize;
+    protected Map<Long, Long> availableColumnsWithTable;
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/QueryDataPoint.java:56]: Can this be final?

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/QueryDataPointImpl.java [hunk 5]
```diff
@@ -49,12 +50,15 @@
...
     private boolean isSubQuery;
+    protected boolean isCommitted;
     private int rowCount;
     private List<String> fieldNames;
     private List<Long> accessedPartitions;
     protected String algCompareString;
     protected String queryClass;
     protected String physicalQueryClass;
+    protected Integer indexSize;
+    protected Map<Long, Long> availableColumnsWithTable;
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/QueryDataPointImpl.java:61]: Can this be final?

File: core/src/main/java/org/polypheny/db/monitoring/events/StatementEvent.java [hunk 6]
```diff
@@ -52,11 +52,16 @@
     protected int rowCount;
     protected boolean isAnalyze;
     protected boolean isSubQuery;
+    protected boolean isCommitted;
     protected String durations;
     protected Map<Long, Set<Long>> accessedPartitions;
     protected LogicalQueryInformation logicalQueryInformation;
     protected String algCompareString;
     protected String physicalQueryClass;
+    protected HashMap<Long, List<Object>> changedValues;
```
Comment (vogti) [inline core/src/main/java/org/polypheny/db/monitoring/events/StatementEvent.java:61]: Can this be final?

File: information/build.gradle [hunk 7]
```diff
@@ -10,6 +10,7 @@ dependencies {
 
     implementation group: "org.apache.commons", name: "commons-text", version: commons_text_version  // Apache 2.0
     implementation group: "org.apache.commons", name: "commons-collections4", version: commons_collections_version  // Apache 2.0
+    api group: "com.google.guava", name: "guava", version: guava_version
```
Comment (vogti) [inline information/build.gradle:13]: Why is this necessary?

File: webui/src/main/java/org/polypheny/db/webui/crud/StatisticCrud.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        Long schemaId;
+
+        try {
+            schemaId = Catalog.getInstance().getSchema( 1, request.tableId.split( "\\." )[0] ).id;
+            tableId = Catalog.getInstance().getTable( schemaId, request.tableId.split( "\\." )[1] ).id;
+
+            StatisticTable test = (StatisticTable) statisticsManager.getTableStatistic( schemaId, tableId );
+            ctx.json( test );
+        } catch ( UnknownTableException | UnknownSchemaException e ) {
+            e.printStackTrace();
```
Comment (vogti) [inline webui/src/main/java/org/polypheny/db/webui/crud/StatisticCrud.java:96]: Return error

File: monitoring/src/main/java/org/polypheny/db/monitoring/statistics/StatisticsManagerImpl.java [hunk 9]
```diff
@@ -0,0 +1,1003 @@
...
+    @Getter
+    private String revalId = null;
+
+    @Getter
+    public volatile ConcurrentHashMap<Long, HashMap<Long, HashMap<Long, StatisticColumn<T>>>> statisticSchemaMap;
+
+    @Getter
+    public ConcurrentHashMap<Long, StatisticTable<T>> tableStatistic;
+
+    //new Additions for additional Information
```
Comment (vogti) [inline monitoring/src/main/java/org/polypheny/db/monitoring/statistics/StatisticsManagerImpl.java:108]: This does not really help :D

File: dbms/src/main/java/org/polypheny/db/processing/AbstractQueryProcessor.java [hunk 10]
```diff
@@ -1400,7 +1399,7 @@ private void monitorResult( ProposedRoutingPlan selectedPlan ) {
      *
      * @param eventData monitoring data to be updated
      */
-    private void finalizeAccessedPartitions( StatementEvent eventData ) {
+    public void finalizeAccessedPartitions( StatementEvent eventData ) {
```
Comment (vogti) [inline dbms/src/main/java/org/polypheny/db/processing/AbstractQueryProcessor.java:1402]: Why changing the visibility to public? This method is only called from within this class.

Comment (isabelge) [reply dbms/src/main/java/org/polypheny/db/processing/AbstractQueryProcessor.java]: Changed back to private.

File: core/src/main/java/org/polypheny/db/StatisticsManager.java (hunk 1 above)
Comment (isabelge) [reply core/src/main/java/org/polypheny/db/StatisticsManager.java]: Changed name to statisticsManager.

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/DdlEvent.java (hunk 2 above)
Comment (isabelge) [reply monitoring/src/main/java/org/polypheny/db/monitoring/events/DdlEvent.java]: Deleted the eventType.

File: information/build.gradle (hunk 7 above)
Comment (isabelge) [reply information/build.gradle:13]: This is used for the EvictingQueue to limit the debug information of the catalog.

File: monitoring/src/main/java/org/polypheny/db/monitoring/statistics/StatisticsManagerImpl.java (hunk 9 above)
Comment (isabelge) [reply monitoring/src/main/java/org/polypheny/db/monitoring/statistics/StatisticsManagerImpl.java]: Changed to Map.

File: webui/src/main/java/org/polypheny/db/webui/crud/StatisticCrud.java (hunk 8 above)
Comment (isabelge) [reply webui/src/main/java/org/polypheny/db/webui/crud/StatisticCrud.java]: Now returns an error.

File: core/src/main/java/org/polypheny/db/monitoring/events/StatementEvent.java (hunk 6 above)
Comment (isabelge) [reply core/src/main/java/org/polypheny/db/monitoring/events/StatementEvent.java]: HashMap is now final.

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/QueryDataPointImpl.java (hunk 5 above)
Comment (isabelge) [reply monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/QueryDataPointImpl.java]: Is now final.

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/QueryDataPoint.java (hunk 4 above)
Comment (isabelge) [reply monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/QueryDataPoint.java:56]: Yes, it is final now.

File: monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/DmlDataPoint.java (hunk 3 above)
Comment (isabelge) [reply monitoring/src/main/java/org/polypheny/db/monitoring/events/metrics/DmlDataPoint.java]: Yes, it is final now.

File: monitoring/src/main/java/org/polypheny/db/monitoring/statistics/StatisticsManagerImpl.java (hunk 9 above)
Comment (vogti) [reply monitoring/src/main/java/org/polypheny/db/monitoring/statistics/StatisticsManagerImpl.java]: I've meant the comment ;)

**Review by vogti — APPROVED** [review]
Thx, @isabelge, for this PR!
